File: python/inql/menu/context_menu.py
# coding: utf-8

from burp.api.montoya.ui.contextmenu import ContextMenuItemsProvider

from java.awt.event import ActionListener, MouseAdapter, MouseEvent
from javax.swing import JMenuItem, JPopupMenu

from ..globals import app, montoya
from ..logger import log
from ..utils.graphql import is_query
from ..utils.ui import visual_error


class SendFromInQL(MouseAdapter):
    def __init__(self, request, include_scanner=False):
        log.debug("Attaching new context menu handler")
        self.request = request

        self.popup = JPopupMenu()
        self.popup.add(SendToIntruder(request))
        self.popup.add(SendToRepeater(request))
        self.popup.addSeparator()
        if include_scanner:
            self.popup.add(ScannerMenuItem(request))
        self.popup.add(AttackerMenuItem(request))
        self.popup.add(GraphiqlMenuItem(request))

    def mousePressed(self, event):
        log.debug("Showing the right click menu")
        if event.button == MouseEvent.BUTTON3:
            self.popup.show(event.getComponent(), event.getX(), event.getY())




class ContextMenuItem(ActionListener, JMenuItem):
    """Action Listener - listens for clicks inside the context menu."""

    # The menu item text
    description = ""

    def __init__(self, request):
        log.debug("Context menu item initialized")
        super(ContextMenuItem, self).__init__(self.description)

        self.request = request
        self.addActionListener(self)

    @property
    def headers(self):
        log.debug("Getting headers")
        log.debug("Request: %s", self.request)
        log.debug("Headers: %s", self.request.headers())
        log.debug("[Headers] %s", [(header.name(), header.value()) for header in self.request.headers()])
        return [(header.name(), header.value()) for header in self.request.headers()
                    if header.name()[0] != ":"]

    @property
    def httpservice(self):
        return self.request.httpService()

    @property
    def url(self):
        return self.request.url()

    def actionPerformed(self, event):
        """Called when a menu item gets clicked."""
        log.debug("Menu item click handler fired: %s", event)
        # TODO: Make sure request is GraphQL before firing event
        self.handler(event)

    # This has to be defined in subclasses
    def handler(self, _):
        log.debug("works")


class ScannerMenuItem(ContextMenuItem):
    description = "Generate queries with InQL Scanner"

    def handler(self, _):
        app.omnibar.run_from_burp(self.url, self.headers)


class AttackerMenuItem(ContextMenuItem):
    description = "Batch attack with InQL Attacker"

    def handler(self, _):
        log.debug("fired the handler")
        app.attacker.send_to(self.url, self.request)
        log.debug("successfully fired the handler")


class GraphiqlMenuItem(ContextMenuItem):
    description = "Open in GraphiQL (embedded web IDE)"

    def handler(self, _):
        visual_error("GraphiQL support not implemented yet.")

class SendToIntruder(ContextMenuItem):
    description = "Send to Intruder"

    def handler(self, _):
        montoya.intruder().sendToIntruder(self.request)


class SendToRepeater(ContextMenuItem):
    description = "Send to Repeater"

    def handler(self, _):
        montoya.repeater().sendToRepeater(self.request)



class ContextMenu(ContextMenuItemsProvider):
    """Global context menu that appears on right-click in various Burp Suite locations."""
    menu_entries = (
        ScannerMenuItem,
        AttackerMenuItem,
        GraphiqlMenuItem
    )

    # TODO: Right now each menu item gets its own request. Is it pointer or a full copy?
    def provideMenuItems(self, ctx):
        """
        Invoked by Burp Suite when the user requests a context menu anywhere in the user interface.

        ctx: ContextMenuEvent, https://portswigger.github.io/burp-extensions-montoya-api/javadoc/burp/api/montoya/ui/contextmenu/ContextMenuEvent.html
        Returns: A list of custom menu items (which may include sub-menus, checkbox menu items, etc.) that should be displayed.
        """
        log.debug("provideMenuItems fired")
        request = self.__rr_from_ctx(ctx)
        log.debug("Context menu successfully identified the selected request")
        if not request:
            return []

        if not is_query(request.bodyToString()):
            log.debug("Context menu not added, because couldn't detect GraphQL content in request")
            return []

        return [entry(request) for entry in self.menu_entries]


    def __rr_from_ctx(self, ctx):
        # Right now we don't show menu if more than one request / HTTP message / scan issue has been selected
        invocation = ctx.invocationType()

        if invocation.containsScanIssues():
            log.debug("Context menu invocation with scan issues")
            issues = ctx.selectedIssues()
            log.debug("there are %s issues", len(issues))
            if len(issues) == 1:
                rrs = issues[0].requestResponses()
                if rrs:
                    log.debug("There are %s rrs", len(rrs))
                    return rrs[0].request()
                log.warn("Couldn't extract request from the selected issue.")
            else:
                log.debug("Not showing the menu, because %s issues have been selected instead of 1", len(issues))

        if invocation.containsHttpRequestResponses():
            log.debug("Context menu invocation with HTTP request-responses")
            rrs = ctx.selectedRequestResponses()
            if len(rrs) == 1:
                log.debug("Found the selected request response")
                return rrs[0].request()
            log.debug("Not showing the menu, because %s request-responses have been selected instead of 1", len(rrs))

        if invocation.containsHttpMessage():
            log.debug("Context menu invocation with HTTP messages")
            # Note that the following API returns java.util.Optional
            messages = ctx.messageEditorRequestResponse().orElse(None)
            if not messages:
                log.debug("Not showing the menu, because no HTTP messages have been selected")
            else:
                log.debug("A HTTP message was selected, nice")
                return messages.requestResponse().request()

        return None

File: src/main/kotlin/inql/ui/ContextMenu.kt
package inql.ui

import burp.Burp
import burp.api.montoya.http.message.requests.HttpRequest
import burp.api.montoya.http.message.responses.HttpResponse
import burp.api.montoya.ui.contextmenu.AuditIssueContextMenuEvent
import burp.api.montoya.ui.contextmenu.ContextMenuEvent
import burp.api.montoya.ui.contextmenu.ContextMenuItemsProvider
import burp.api.montoya.ui.contextmenu.InvocationType
import burp.api.montoya.ui.contextmenu.MessageEditorHttpRequestResponse
import inql.graphql.GraphQLRequestTransformException
import inql.graphql.GraphQLRequestTransformer
import inql.graphql.GraphQLTransportFormat
import inql.history.HistoryHostKey
import inql.history.HistoryTracker
import inql.Config
import inql.InQL
import inql.Logger
import inql.externaltools.ExternalToolsService
import inql.externaltools.ExternalToolsService.Companion.sendRequestToEmbeddedTool
import inql.schema.corrections.GraphQLTypeSuggestionParser
import javax.swing.SwingUtilities
import java.awt.Component
import java.awt.Toolkit
import java.awt.event.ActionEvent
import java.awt.event.KeyEvent
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import javax.swing.*

open class MenuAction(val name: String, val keyStroke: KeyStroke?, val action: (ActionEvent) -> Unit) :
    AbstractAction(name) {
    override fun actionPerformed(e: ActionEvent) {
        this.action(e)
    }
}

/*
    This class provides the Context Menu that is opened when the user Right-Clicks inside InQL (e.g. in the Scanner Tab Results
    The actions have associated Keyboard Shortcuts so that standard Burp shortcuts can be used from InQL

    This class is also extended below from SendToInqlHandler which instead provides the Extension Context Menu "InQL >" for
    Burp's standard context menu in other Burp tools (Scanner, Proxy, etc).
 */
abstract class SendFromInqlHandler(val inql: InQL, val includeInqlScanner: Boolean = false, val includeSaveToFile: Boolean = true) : MouseAdapter() {
    private val popup = JPopupMenu()

    // ===== Actions associated with Menu Items

    protected val sendToIntruderAction = MenuAction(
        "Send to Intruder",
        KeyStroke.getKeyStroke(KeyEvent.VK_I, Toolkit.getDefaultToolkit().menuShortcutKeyMaskEx),
    ) {
        this.sendRequestToIntruder()
    }
    protected val sendToRepeaterAction = MenuAction(
        "Send to Repeater",
        KeyStroke.getKeyStroke(KeyEvent.VK_R, Toolkit.getDefaultToolkit().menuShortcutKeyMaskEx),
    ) {
        this.sendRequestToRepeater()
    }
    protected val sendToInqlScannerAction = MenuAction("Analyse in Introspection Scanner", null) {
        this.sendRequestToInqlScanner()
    }
    protected val sendToInqlAttackerAction = MenuAction("Open in Batch Queries", null) {
        this.sendRequestToInqlAttacker()
    }
    protected val sendToInqlFingerprinterAction = MenuAction("Open in Engine Fingerprinter", null) {
        this.sendRequestToInqlFingerprinter()
    }
    protected val sendToGraphiqlAction = MenuAction("Open in GraphiQL", null) {
        this.sendRequestToGraphiQL()
    }
    protected val sendToVoyagerAction = MenuAction("Open in GraphQL Voyager", null) {
        this.sendRequestToVoyager()
    }

    protected val saveToFileAction = MenuAction("Save to file", null) {
        val filechooser = JFileChooser()
        if (filechooser.showSaveDialog(Burp.Montoya.userInterface().swingUtils().suiteFrame()) == JFileChooser.APPROVE_OPTION) {
            val file = filechooser.selectedFile
            file.writeText(this.getText())
        }
    }

    // A list of optional actions to include based on config values:
    //         "integrations.graphiql" to true,
    //         "integrations.voyager" to true,
    // Check those with: Config.getInstance().getBoolean("integrations.graphiql")
    protected val sendToEmbeddedToolActions = mutableListOf<MenuAction>().apply {
        val config = Config.getInstance()
        if (config.getBoolean("integrations.graphiql") == true) {
            this.add(sendToGraphiqlAction)
        }
        if (config.getBoolean("integrations.voyager") == true) {
            this.add(sendToVoyagerAction)
        }
    }

    /* The following list is currently used for:
     - Enable/Disable (grey-out) context menu items when the user
        right-clicks something that is not actually a GraphQL item, e.g. a point of interest in the scanner results.
     - Provide Keyboard Shortcuts (CTRL+R, CTRL+I, etc)
     */
    protected val sendFromInqlActions = mutableListOf<MenuAction>(
        sendToIntruderAction,
        sendToRepeaterAction,
        sendToInqlAttackerAction,
        sendToInqlFingerprinterAction,
        sendToInqlScannerAction,
        sendToGraphiqlAction,
        sendToVoyagerAction
    )
    abstract fun getRequest(): HttpRequest?
    abstract fun getText(): String
    override fun mousePressed(e: MouseEvent) {
        if (e.button == MouseEvent.BUTTON3) {    // Right Click only
            this.setContextActions()
            this.popup.show(e.component, e.x, e.y)
        }
    }

    /** Subclasses may add items (e.g. Copy) after the standard entries. */
    protected open fun appendBottomContextMenuItems(popup: JPopupMenu) {}

    // Populate the right click menu in InQL views (InQL Scanner, GraphQL editor view)
    // The context menus added by Burp itself **are not handled here** (e.g. Repeater - Raw editor - right click)
    // In order to add elements to Burp's menu (Extensions - InQL - ...), modify sendToInqlComponents in SendToInqlHandler class
    private fun setContextActions() {
        this.popup.removeAll()

        if (this.popup.componentCount > 0) {
            this.popup.addSeparator()
        }

        this.popup.add(this.sendToIntruderAction)
        this.popup.add(this.sendToRepeaterAction)
        this.popup.addSeparator()

        if (this.includeInqlScanner) {
            this.popup.add(this.sendToInqlScannerAction)
        }
        this.popup.add(this.sendToInqlAttackerAction)
        this.popup.add(this.sendToInqlFingerprinterAction)

        val embeddedActions = this.sendToEmbeddedToolActions
        if (embeddedActions.isNotEmpty()) {
            this.popup.addSeparator()

            for (action in embeddedActions) {
                this.popup.add(action)
            }
        }

        if (this.includeSaveToFile) {
            this.popup.addSeparator()
            this.popup.add(this.saveToFileAction)
        }

        appendBottomContextMenuItems(this.popup)
    }

    // ===== Convenience methods for the actions
    private fun sendRequestToIntruder() {
        Burp.Montoya.intruder().sendToIntruder(this.getRequest() ?: return)
    }

    private fun sendRequestToRepeater() {
        Burp.Montoya.repeater().sendToRepeater(this.getRequest() ?: return)
    }

    private fun sendRequestToInqlScanner() {
        inql.scanner.newTabFromRequest(this.getRequest() ?: return)
    }

    private fun sendRequestToInqlAttacker() {
        inql.attacker.loadFromRequest(this.getRequest() ?: return)
    }

    private fun sendRequestToInqlFingerprinter() {
        inql.fingerprinter.loadFromRequest(this.getRequest() ?: return)
    }

    private fun sendRequestToGraphiQL() {
        Logger.debug("Send Request to GraphiQL")
        sendRequestToEmbeddedTool(this.getRequest(), ExternalToolsService.Companion.TOOL.TOOL_GRAPHIQL)
    }

    private fun sendRequestToVoyager() {
        Logger.debug("Send Request to Voyager")
        sendRequestToEmbeddedTool(this.getRequest(), ExternalToolsService.Companion.TOOL.TOOL_VOYAGER)
    }

    fun setEnabled(enabled: Boolean) {
        this.sendFromInqlActions.forEach {
            it.isEnabled = enabled
        }
    }

    fun addRightClickHandler(c: Component) {
        c.addMouseListener(this)
    }

    fun addKeyboardShortcutHandler(c: JComponent) {
        for (action in this.sendFromInqlActions) {
            if (action.keyStroke == null) continue
            c.inputMap.put(action.keyStroke, action.name)
            c.actionMap.put(action.name, action)
        }
    }
}

class SendToInqlHandler(inql: InQL) : SendFromInqlHandler(inql), ContextMenuItemsProvider {
    class BurpMenuItem(action: MenuAction) : JMenuItem(action.name) {
        init {
            this.addActionListener {
                action.action(it)
            }
        }
    }

    private var request: HttpRequest? = null

    // This only sets Right Click handlers for the Burp's own menus. Menus added by InQL are handled
    // in setContextActions()
    private val extractHistorySchemaAction = MenuAction("Extract Schema from History", null) {
        this.extractGraphQLSchemaFromHost()
    }

    private fun typeRenameSubmenu(actions: List<MenuAction>): JMenu? {
        if (actions.isEmpty()) return null
        return JMenu("Add type correction").also { submenu ->
            for (action in actions) {
                submenu.add(BurpMenuItem(action))
            }
        }
    }

    private fun transformRequestSubmenu(
        editor: MessageEditorHttpRequestResponse,
    ): JMenu? {
        val request = editor.requestResponse().request()
        if (GraphQLRequestTransformer.parsePayload(request) == null) return null

        return JMenu("Transform Request to:").also { submenu ->
            for (format in GraphQLTransportFormat.entries.filter { it.showInTransformMenu() }) {
                submenu.add(
                    BurpMenuItem(
                        MenuAction(format.menuLabel!!, null) {
                            transformRequestInEditor(editor, format)
                        },
                    ),
                )
            }
        }
    }

    private fun transformRequestInEditor(
        editor: MessageEditorHttpRequestResponse,
        target: GraphQLTransportFormat,
    ) {
        val request = editor.requestResponse().request()
        try {
            val transformed = GraphQLRequestTransformer.transform(request, target)
            editor.setRequest(transformed)
        } catch (e: GraphQLRequestTransformException) {
            showTransformError(e.message ?: "Request transformation failed.")
        } catch (_: UnsupportedOperationException) {
            showTransformError("This request cannot be edited in the current view.")
        } catch (e: Exception) {
            showTransformError("Request transformation failed: ${e.message}")
        }
    }

    private fun showTransformError(message: String) {
        SwingUtilities.invokeLater {
            JOptionPane.showMessageDialog(
                Burp.Montoya.userInterface().swingUtils().suiteFrame(),
                message,
                "InQL",
                JOptionPane.WARNING_MESSAGE,
            )
        }
    }

    private fun sendToInqlComponents(
        typeRenameActions: List<MenuAction> = emptyList(),
        messageEditor: MessageEditorHttpRequestResponse? = null,
    ): MutableList<Component> {
        return mutableListOf<Component>(
            BurpMenuItem(super.sendToInqlScannerAction),
            BurpMenuItem(super.sendToInqlAttackerAction),
            BurpMenuItem(super.sendToInqlFingerprinterAction),
        ).apply {
            for (action in super.sendToEmbeddedToolActions) {
                this.add(BurpMenuItem(action))
            }
            if (this@SendToInqlHandler.selectedHost != null) {
                this.add(BurpMenuItem(extractHistorySchemaAction))
            }
            typeRenameSubmenu(typeRenameActions)?.let { add(it) }
            messageEditor?.let { editor ->
                transformRequestSubmenu(editor)?.let { submenu ->
                    add(JSeparator())
                    add(submenu)
                }
            }
        }
    }

    private var selectedHost: String? = null

    private fun extractGraphQLSchemaFromHost() {
        val host = this.selectedHost ?: return
        if (!HistoryTracker.isRunning()) {
            HistoryTracker.start(inql)
        }
        HistoryTracker.extractSchemaForHost(host)
    }

    private fun hostFromContext(event: ContextMenuEvent): String? {
        if (event.invocationType() == InvocationType.SITE_MAP_TREE) {
            val requestResponses = event.selectedRequestResponses()
            if (requestResponses.isNotEmpty()) {
                return HistoryHostKey.fromRequest(requestResponses[0].request())
            }
        }
        val request = requestFromContext(event)
        if (request != null) {
            return HistoryHostKey.fromRequest(request)
        }
        return null
    }

    private fun requestFromContext(event: ContextMenuEvent): HttpRequest? {
        val invocationType = event.invocationType()
        if (invocationType.containsHttpRequestResponses()) {
            val requestResponses = event.selectedRequestResponses()
            if (requestResponses.size != 1) return null
            return requestResponses[0].request()
        }
        if (invocationType.containsHttpMessage()) {
            val msg = event.messageEditorRequestResponse().orElse(null) ?: return null
            return msg.requestResponse().request()
        }
        return null
    }

    private fun responseFromContext(event: ContextMenuEvent): HttpResponse? {
        val invocationType = event.invocationType()
        if (invocationType.containsHttpRequestResponses()) {
            val requestResponses = event.selectedRequestResponses()
            if (requestResponses.size != 1) return null
            return requestResponses[0].response()
        }
        if (invocationType.containsHttpMessage()) {
            val msg = event.messageEditorRequestResponse().orElse(null) ?: return null
            return msg.requestResponse().response()
        }
        return null
    }

    private fun typeRenameActionsFromResponse(
        request: HttpRequest,
        response: HttpResponse,
    ): List<MenuAction> {
        val suggestions = GraphQLTypeSuggestionParser.parseTypeRenameSuggestions(response.bodyToString())
        if (suggestions.isEmpty()) return emptyList()
        return suggestions.map { suggestion ->
            MenuAction(
                "${suggestion.wrongType} → ${suggestion.suggestedType}",
                null,
            ) {
                applyTypeRenameCorrection(request, suggestion.wrongType, suggestion.suggestedType)
            }
        }
    }

    private fun applyTypeRenameCorrection(request: HttpRequest, wrongType: String, suggestedType: String) {
        val host = HistoryHostKey.fromRequest(request)
        val applied = inql.scanner.applyTypeRenameCorrection(request, host, wrongType, suggestedType)
        if (applied) return
        SwingUtilities.invokeLater {
            JOptionPane.showMessageDialog(
                Burp.Montoya.userInterface().swingUtils().suiteFrame(),
                "Could not apply type rename for $host.\n" +
                    "Open or extract a schema for this host in InQL Scanner first.",
                "InQL",
                JOptionPane.WARNING_MESSAGE,
            )
        }
    }

    private fun requestFromIssues(event: AuditIssueContextMenuEvent): HttpRequest? {
        val invocationType = event.invocationType()
        if (invocationType.containsScanIssues()) {
            val issues = event.selectedIssues()
            if (issues.size != 1) return null
            val requestResponses = issues[0].requestResponses()
            if (requestResponses.isEmpty()) return null
            return requestResponses[0].request()
        }
        return null
    }

    override fun provideMenuItems(event: ContextMenuEvent): MutableList<Component>? {
        this.selectedHost = this.hostFromContext(event)
        if (event.invocationType() == InvocationType.SITE_MAP_TREE && this.selectedHost != null) {
            return mutableListOf(BurpMenuItem(extractHistorySchemaAction))
        }
        val request = this.requestFromContext(event) ?: return null
        this.request = request
        val typeRenameActions = responseFromContext(event)?.let { response ->
            typeRenameActionsFromResponse(request, response)
        } ?: emptyList()
        val messageEditor = event.messageEditorRequestResponse()
            .filter { it.selectionContext() == MessageEditorHttpRequestResponse.SelectionContext.REQUEST }
            .orElse(null)
        return this.sendToInqlComponents(typeRenameActions, messageEditor)
    }

    override fun provideMenuItems(event: AuditIssueContextMenuEvent?): MutableList<Component> {
        if (event == null) return mutableListOf()
        this.selectedHost = null
        this.request = this.requestFromIssues(event) ?: return mutableListOf()
        return this.sendToInqlComponents()
    }

    override fun getRequest(): HttpRequest? {
        return this.request
    }

    override fun getText(): String {
        return this.request.toString()
    }
}
